### PR TITLE
Redesign and completed authentication backend. Add signupverification.html.

### PR DESCRIPTION
**Pull request by saraholiviawu**
**Notable changes since last Authentication commit**: 
  - **Generalize functions to reduce code repetition and decrease calls to read/write to Firebase**: A few major examples include:
     -  The function that presents a Google sign-in pop up was generalized to handle both the 'Log in' and 'Sign up' functionality, depending on the parameter passed in. 
     - Two functions that both got a document from Firestore based on uid were generalized to take in a user id and lambda function. 
     - The function to sign out a user, which is called on all pages, was generalized to sign out a user without redirecting to the Landing page if a parameter is passed in. Signing out is necessary in the authentication pages because generating the Google pop up automatically authenticates a user for the local session. In some cases, however, the user should not yet exist or be registered because he/she attempted signing in using the wrong method or has not yet verified their email. 
  - **Store user type in local storage**: When a new user selects either of the buttons 'I'm a community member' or 'I'm a business owner,' their choice is stored in local storage. When the user signs up on the following page, the local storage is accessed to retrieve this value and written to Firebase Firestore alongside the user's name, user id, and email. 
  - **Store user id, userName, isBusinessOwner in local storage**: Only occurs when user signs in. These local storage values deleted when user signs out. Will reduce Firestore reads significantly.
  - **Email Verification**: Add signupverification.html. After a new user signs up, a verification link is sent to their email and simultaneously redirect this page, which presents text instructing user to check their email. Once the user clicks the email verification link, they are considered verified and are redirected to log in for the first time. Note that Google sign-in automatically is considered verified. This functionality will be particularly important when implementing other sign-in types such as Facebook.
  - **Prevent an existing user from signing up**: A check is placed to see if the user exists in Firebase Firestore. If so, an error message is displayed and the user is signed out of the existing session.
  - **Prevent a new user from logging in**: A check is placed to see if the user exists in Firebase Firestore. If not, an error message is displayed and the user is signed out of the existing session. 
  - **Prevent an unverified user from logging in**: A check is placed to see if the user has verified their email. If not, an error message is displayed and the user is signed out of the existing session.

**Demo**
1. New user attempts to log in and receives error. New user then chooses to sign up, selects a user type, and authenticates through Google. User is redirected to signupverification.html. User clicks verification email, which directs them back to login page. After logging in, user is redirected to the landing page.

New user sign-up flow
See full video: https://drive.google.com/file/d/1TYjfQ3z3Idu15tly5cebPD4tcSQGvdGX/view

2. Existing user attempts to create a new account.

Existing user attempts to create a new account
See full video: https://drive.google.com/file/d/1BDv8pJmWMMJ62yegTnrclaAXy5ZUvNP4/view

3.  New user that has already created an account attempts to log in without yet verifying email. Note a Google sign in is always considered verified and the user may log in without going through this step. This feature will be useful for other sign in methods.
Email not yet verified
 See full video: https://drive.google.com/file/d/1ENCUiqUR9YcVHa8CaLCzvGW4emzqKtGc/view

### CODE REVIEW

File: ripple/src/main/webapp/posts.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
+/* Update function to fetch all comments for a particular post */
+function viewAllPostComments() {
+  alert("Fetch all comments for this post!");
+}
+
+/* Display an alert if user presses enter to comment on a post */
+function addComment(e) {
+  comment = document.getElementById("add-comment").value;
+  if (e.keyCode === 13) {
```
Comment (peter14f) [inline ripple/src/main/webapp/posts.js:9]: There is another place checking for value being 13.
Can we make a global variable ENTER_KEY and use it.

File: ripple/src/main/webapp/script.js [hunk 2]
```diff
@@ -22,17 +29,65 @@ function fetchBlobstoreUploadUrl(formId, fileId, webUrl) {
...
+function getBlobKey(uid, blobKey) {
+  var newBlobKey = getParameterByName('blob-key');
+  console.log("Parameter blobKey: " + newBlobKey);
+  // New image uploaded. Store blobkey in firestore.
+  if (newBlobKey != null) {
+    console.log("New image. Blobkey parameter: ", newBlobKey);
+    db.collection("users").doc(uid).update({
+      blobKey: newBlobKey,
+    })
+        .then(function() {
```
Comment (peter14f) [inline ripple/src/main/webapp/script.js:42]: then() and catch() formatting seems off?

File: ripple/src/main/webapp/script.js [hunk 3]
```diff
@@ -22,17 +29,65 @@ function fetchBlobstoreUploadUrl(formId, fileId, webUrl) {
...
+    serveBlob(blobKey, "nav-bar-avatar");
+    serveBlob(blobKey, "profile-image");
+  }
+}
+
+/* Read parameter in URL of window */
+function getParameterByName(name) {
+  console.log('called getParameterByName()');
+  url = window.location.href;
+  name = name.replace(/[\[\]]/g, '\\$&');
```
Comment (peter14f) [inline ripple/src/main/webapp/script.js:62]: Add a comment here explaining what the regex is trying to accomplish.

File: ripple/src/main/webapp/accountsettings.html [hunk 4]
```diff
@@ -16,16 +16,33 @@
     <script src="https://www.gstatic.com/firebasejs/7.15.5/firebase-auth.js"></script>
     <script src="https://www.gstatic.com/firebasejs/7.15.5/firebase-firestore.js"></script>
     <script>
-      firebase.initializeApp(config);
+      var apiKey = config.API_KEY;
```
Comment (smruthibalajee) [inline ripple/src/main/webapp/accountsettings.html:19]: Will this config setup be moved to the config file and called in one line for shorter code?

File: ripple/src/main/webapp/posts.js (hunk 1 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/posts.js:9]: Fixed

File: ripple/src/main/webapp/script.js (hunk 2 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/script.js:42]: Fixed

File: ripple/src/main/webapp/script.js (hunk 3 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/script.js:62]: Fixed

File: ripple/src/main/webapp/accountsettings.html (hunk 4 above)
Comment (saraholiviawu) [reply ripple/src/main/webapp/accountsettings.html:19]: Fixed

**Pull request merged**